SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -0,0 +1,202 @@
+	{
+		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Log, TEXT("Authority over the VirtualWorkerTranslationManager has changed. This worker %s authority."), bAuthoritative ? TEXT("now has") : TEXT("does not have"));
+
+		if (bAuthoritative)
+		{
+			// TODO(zoning): The prototype had an unassigned workers list. Need to follow up with Tim/Chris about whether
+			// that is necessary or we can continue to use the (possibly) stale list until we receive the query response.
+
+			// Query for all connection entities, so we can detect if some worker has died and needs to be updated in
+			// the mapping.
+			QueryForWorkerEntities();
+		}
+	}
+}
+
+// For each entry in the map, write a VirtualWorkerMapping type object to the Schema object.
+void SpatialVirtualWorkerTranslationManager::WriteMappingToSchema(Schema_Object* Object) const
+{
+	for (auto& Entry : VirtualToPhysicalWorkerMapping)
+	{
+		Schema_Object* EntryObject = Schema_AddObject(Object, SpatialConstants::VIRTUAL_WORKER_TRANSLATION_MAPPING_ID);
+		Schema_AddUint32(EntryObject, SpatialConstants::MAPPING_VIRTUAL_WORKER_ID, Entry.Key);
+		SpatialGDK::AddStringToSchema(EntryObject, SpatialConstants::MAPPING_PHYSICAL_WORKER_NAME, Entry.Value);
+	}
+}
+
+// This method is called on the worker who is authoritative over the translation mapping. Based on the results of the
+// system entity query, assign the VirtualWorkerIds to the workers represented by the system entities.
+void SpatialVirtualWorkerTranslationManager::ConstructVirtualWorkerMappingFromQueryResponse(const Worker_EntityQueryResponseOp& Op)
+{
+	// The query response is an array of entities. Each of these represents a worker.
+	for (uint32_t i = 0; i < Op.result_count; ++i)
+	{
+		const Worker_Entity& Entity = Op.results[i];
+		for (uint32_t j = 0; j < Entity.component_count; j++)
+		{
+			const Worker_ComponentData& Data = Entity.components[j];
+			// System entities which represent workers have a component on them which specifies the SpatialOS worker ID,
+			// which is the string we use to refer to them as a physical worker ID.
+			if (Data.component_id == SpatialConstants::WORKER_COMPONENT_ID)
+			{
+				const Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data.schema_type);
+
+				const FString& WorkerType = SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::WORKER_TYPE_ID);
+
+				// TODO(zoning): Currently, this only works if server workers never die. Once we want to support replacing
+				// workers, this will need to process UnassignWorker before processing AssignWorker.
+				if (WorkerType.Equals(SpatialConstants::DefaultServerWorkerType.ToString()) &&
+					!UnassignedVirtualWorkers.IsEmpty())
+				{
+					AssignWorker(SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::WORKER_ID_ID));
+				}
+			}
+		}
+	}
+}
+
+// This will be called on the worker authoritative for the translation mapping to push the new version of the map
+// to the spatialOS storage.
```

Comment:
NIT, gotta follow that brand guide 😄 
```suggestion
// to the SpatialOS storage.
```

Reply:
Done